content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
++++
+title = "MemPass"
+extra.author = "Eashan Garg and Sameer Lal"
+extra.bio = """
+Eashan Garg is a senior undergraduate studying CS and English.
+
+Sameer Lal is a CS Master of Engineering student. He studied ECE and Math in his undergraduate.
+"""
++++
+
+While manual access to memory allocation/deallocation can allow the experienced programmer to significantly improve the performance of their program, it can also open the door to a host of memory safety bugs: use after free, double free, out of bounds memory accesses, and the like. Detecting these problems are difficult on the programmer's end and leave room for exploitation by malicious users.
+
+## Memory Safety Vulnerabilities
+
+Many memory safety vulnerabilities occur when the user tries to access or deallocate memory that isn't available for use, either because it was freed, or never allocated in the first place. Take for example the following program:
+
+```c
+int ptr*;
+ptr = (int *)malloc(4 * sizeof(int));
+free(ptr);
+```
+
+After allocating 16 bytes, the starting address for this block of memory is stored in `ptr`, and freed shortly afterward.
+
+Now, assume that the programmer attempts to access the memory stored at the address in `ptr`.
+
+```c
+int *ptr, *ptr2;
+ptr = (int *)malloc(4 * sizeof(int));
+ptr[0] = 4;
+free(ptr);
+ptr2 = (int *)malloc(4 * sizeof(int));
+printf("%d", ptr[0]);
+```
+
+Since `ptr` has been freed, the second call to `malloc` (which is of same size as the first) is likely to allocate the free space `ptr` points to. Now, lets assume a malicious attacker managed to store a dangerous payload in this space. Since the program tries to use `ptr` after it has been freed, it might actually end up accessing the attacker's payload, not the original `ptr` data as intended!
+
+Double free errors occur upon consecutive `free()` call with the same memory address. For instance, consider the following code snippet taken from a Double-Free Toyota ITC test case:

```

Comment:
"Toyota ITC" won't be familiar to most readers, so consider introducing it and linking to some sort of description.